### PR TITLE
CSV and CMVA weights

### PR DESCRIPTION
**Pull request by DylanHsu**
Please do not merge yet

### CODE REVIEW

**Comment by DylanHsu** [comment]
Mostly done putting it in.
Need to test it on a Panda 008 file and validate.

**Comment by DylanHsu** [comment]
Unit test, 1000 WH(bB) events):
http://t3serv001.mit.edu/~dhsu/github/2017-11-21/testskim_whbb_008.root

Event scale factors and associated uncertainties look reasonable.

http://t3serv001.mit.edu/~dhsu/github/2017-11-21/cmvaMinWHbb.png
http://t3serv001.mit.edu/~dhsu/github/2017-11-21/cmvaMaxWHbb.png
http://t3serv001.mit.edu/~dhsu/github/2017-11-21/cmvaLeadingJetNoHbbFound.png

Need to add an enum for the CSV/CMVA weight variations to clean up the internal storage of GeneralTree and the code that fills them.

**Comment by sidnarayanan** [comment]
You might be able to reuse some of this stuff:

https://github.com/sidnarayanan/PandaAnalysis/blob/dev-refactor/Flat/interface/GeneralTree.h#L37-L73

https://github.com/sidnarayanan/PandaAnalysis/blob/dev-refactor/Flat/src/ModulesBTagSFs.cc#L173-L177

**Comment by DylanHsu** [comment]
Please review PR.

Unit test: http://t3serv001.mit.edu/~dhsu/github/2017-11-27/testskim.root

Examples of event weights with the variations:
http://t3serv001.mit.edu/~dhsu/github/2017-11-27/sf_cmvaWeight_cent.pdf
http://t3serv001.mit.edu/~dhsu/github/2017-11-27/sf_cmvaWeight_HFup.pdf
http://t3serv001.mit.edu/~dhsu/github/2017-11-27/sf_cmvaWeight_LFdown.pdf

For simplicity, the variations are not to be multiplied with the central weight, they are their own standalone event weight [(code here) ](https://github.com/DylanHsu/PandaAnalysis/blob/d4d7d0c80ad01c703fd794cbd1b7e7df93e13512/Flat/src/ModulesBTagSFs.cc#L212-L215)

File: Flat/config/GeneralTree.cfg [hunk 1]
```diff
@@ -25,7 +25,41 @@ sf_lepID                   float
 sf_lepIso                  float
 sf_lepTrack                float
 sf_pho                     float
-# btag SFs - these are done differently now
+# btag SFs
+sf_cmvaweight            float
```
Comment (sidnarayanan) [inline Flat/config/GeneralTree.cfg]: Why are these booked as individual floats?

File: Flat/src/GeneralTree.cc [hunk 2]
```diff
@@ -34,6 +34,11 @@ GeneralTree::GeneralTree() {
       }
     }
   }
+  
+  for (unsigned iShift=0; iShift<17; iShift++) {
```
Comment (sidnarayanan) [inline Flat/src/GeneralTree.cc]: Can we avoid hardcoding 17?

File: Flat/src/ModulesBTagSFs.cc [hunk 3]
```diff
@@ -179,3 +179,40 @@ void PandaAnalyzer::JetBtagSFs()
...
+void PandaAnalyzer::JetCMVAWeights() 
+{
+  for (unsigned iShift=0; iShift<17; iShift++) {
+    GeneralTree::csvShift theShift = gt->csvShifts[iShift];
+    gt->sf_csvWeights[theShift] = 1;
+  }
+  if (centralJets.size() < 1) return;
+
+  //get vectors of jet properties
+  std::vector<double> jetPts, jetEtas, jetCSVs, jetCMVAs;
```
Comment (sidnarayanan) [inline Flat/src/ModulesBTagSFs.cc:191]: Can you reserve them to `centralJets.size()` ?

File: Flat/src/ModulesJets.cc [hunk 4]
```diff
@@ -70,7 +70,7 @@ void PandaAnalyzer::JetBasics()
     }
 
 
-    if (jet.pt()>30) { // nominal jets
+    if (jet.pt()>20) { // nominal jets
```
Comment (sidnarayanan) [inline Flat/src/ModulesJets.cc]: No. We cannot mess up other analyses like this. Make it configurable and only for h->bb.

File: Flat/src/ModulesJets.cc [hunk 5]
```diff
@@ -349,7 +354,7 @@ void PandaAnalyzer::JetHbbReco()
   if (centralJets.size() > 1) {
     vector<Jet*> csvSortedJets = centralJets;
     sort(csvSortedJets.begin(), csvSortedJets.end(),
```
Comment (sidnarayanan) [inline Flat/src/ModulesJets.cc]: Rename csvSorted to btagSorted to avoid confusion.

File: Flat/src/ModulesJets.cc [hunk 6]
```diff
@@ -349,7 +354,7 @@ void PandaAnalyzer::JetHbbReco()
...
-        [](panda::Jet *x, panda::Jet *y) -> bool { return x->csv > y->csv; });
+        [](panda::Jet *x, panda::Jet *y) -> bool { return x->cmva > y->cmva; });
```
Comment (sidnarayanan) [inline Flat/src/ModulesJets.cc]: Again, no. Hardcoding changes like this messes up other analyses. Make it configurable, only for h->bb.

File: Flat/config/GeneralTree.cfg (hunk 1 above)
Comment (DylanHsu) [reply Flat/config/GeneralTree.cfg]: did not edit the config with the enum changes

File: Flat/src/GeneralTree.cc (hunk 2 above)
Comment (DylanHsu) [reply Flat/src/GeneralTree.cc]: maybe

File: Flat/src/ModulesBTagSFs.cc (hunk 3 above)
Comment (DylanHsu) [reply Flat/src/ModulesBTagSFs.cc:191]: yeah

File: Flat/src/ModulesJets.cc (hunk 4 above)
Comment (DylanHsu) [reply Flat/src/ModulesJets.cc]: Oops, makes sense! ( We also use 30 GeV for ZH)

File: Flat/src/ModulesJets.cc (hunk 6 above)
Comment (DylanHsu) [reply Flat/src/ModulesJets.cc]: Ok, already have "useCMVA" flag, just need to add it here

File: Flat/src/GeneralTree.cc (hunk 2 above)
Comment (DylanHsu) [reply Flat/src/GeneralTree.cc]: Answer is no because the enums are not sequential integers. But we can hardcode 17 in the header

**Comment by DylanHsu** [comment]
comments addressed, please see latest commits

File: Flat/src/ModulesJets.cc [hunk 7]
```diff
@@ -102,7 +104,7 @@ void PandaAnalyzer::JetBasics()
       if (analysis->vbf)
         JetVBFBasics(jet);
 
-      if (analysis->monoh) {
+      if (analysis->monoh | analysis->hbb) {
```
Comment (sidnarayanan) [inline Flat/src/ModulesJets.cc]: `||` not `|`

**Comment by sidnarayanan** [comment]
Happy to merge once typo is fixed

File: Flat/src/ModulesJets.cc (hunk 7 above)
Comment (DylanHsu) [reply Flat/src/ModulesJets.cc]: """fixed"""

**Pull request merged**